# B5/B6/B7 — Review variant_claims candidates, guarantee floor conflicts, measure extraction quality

Per `docs/TODO-stage4.md` Track B:
- **B5**: review every row in `variant_claims_candidates.json` against its source
  `passage_ref`; promote approved rows by setting `trust_tier=1`.
- **B6**: guarantee the seed contains all three floor conflicts (Aphrodite parentage,
  Io parentage, Achilles death) regardless of what extraction found — hand-add only
  what extraction missed.
- **B7** (diagnostic, non-blocking): measure how many of the two *measurable* floor
  conflicts (Aphrodite, Achilles — Io is structurally excluded, see below) the pipeline
  detected unaided.

Run B2 (`python -m extraction.run_extraction`) first if you haven't — this notebook
reads its output from `extraction/output/variant_claims_candidates.json` and writes
back to the same file (never a second copy — `trust_tier` is the only thing that
changes on review).

In [12]:
import json
import sys
from pathlib import Path

INGESTION_ROOT = Path.cwd().parent
sys.path.insert(0, str(INGESTION_ROOT))

from dotenv import load_dotenv
load_dotenv()

CANDIDATES_PATH = INGESTION_ROOT / "extraction" / "output" / "variant_claims_candidates.json"

In [13]:
from loader.source_registry import SOURCE_REGISTRY
from loader.text_cleaner import clean
from extraction.segmentation import segment

candidates = json.loads(CANDIDATES_PATH.read_text(encoding="utf-8"))
print(f"{len(candidates)} candidate rows loaded")

7429 candidate rows loaded


## B7 — Extraction-quality metric (diagnostic, non-blocking)

Measured against the raw candidates, **before** any B6 hand-add below. Only Aphrodite
and Achilles are measurable this way (`N/2`, not `N/3`) — Io's two variants are both
attributed to Apollodorus (a single `source_id`), so the `count(DISTINCT source_id) >=
2` detector can never emit it; that's expected, not a pipeline miss.

In [14]:
def floor_conflict_detected(subject: str, claim_type: str) -> bool:
    rows = [
        c for c in candidates
        if c["subject_name"].strip().lower() == subject.lower() and c["claim_type"] == claim_type
    ]
    return len({c["source_id"] for c in rows}) >= 2


aphrodite_detected = floor_conflict_detected("Aphrodite", "parentage")
achilles_detected = floor_conflict_detected("Achilles", "death")
detected_count = sum([aphrodite_detected, achilles_detected])

print(f"floor conflicts detected: {detected_count}/2")
print(f"  Aphrodite parentage: {'detected' if aphrodite_detected else 'MISSED'}")
print(f"  Achilles death:      {'detected' if achilles_detected else 'MISSED'}")
print("  Io parentage:        structurally excluded (single-source, not measurable)")

floor conflicts detected: 2/2
  Aphrodite parentage: detected
  Achilles death:      detected
  Io parentage:        structurally excluded (single-source, not measurable)


## B5 — Review every candidate row, grouped by disagreement

Dumping all 7,429 rows in one cell floods Jupyter's IOPub rate limit (and isn't readable anyway). The real review unit is the **841 distinct `(subject, claim_type)` groups** -- each one disagreement, with 1+ candidate rows supporting each side. Review one group at a time with `review_group(subject, claim_type)` below.

For each row shown: check the `claim_value` against the segment text -- the exact text the LLM saw, re-derived the same way `run_extraction.py` builds it (**not** a `narrative_chunks` lookup, which uses different chunk boundaries) -- and decide whether it's a real, correctly-attributed claim.

In [15]:
def build_segment_map(source):
    raw = (INGESTION_ROOT / source.file_path).read_text(encoding="utf-8")
    cleaned = clean(raw)
    segs = segment(cleaned, source.author, source.work, source.passage_ref_extractor)
    return {s.passage_ref: s.text for s in segs}


segment_maps = {s.source_id: build_segment_map(s) for s in SOURCE_REGISTRY}

In [ ]:
from collections import defaultdict

GROUP_INVENTORY_PATH = INGESTION_ROOT / "audit" / "group_inventory_findings.json"

groups: dict[tuple[str, str], list[int]] = defaultdict(list)
for i, c in enumerate(candidates):
    groups[(c["subject_name"].strip().lower(), c["claim_type"])].append(i)

print(f"{len(groups)} distinct (subject, claim_type) groups across {len(candidates)} candidate rows")
print("(one line per group -- pick a subject/claim_type below and call review_group(...) to see its rows)\n")

# Track C6 (DEV-104): ranked by Track B's A8 subject prominence instead of the old flat
# alphabetical dump, so a tranche is picked by prominence + source count rather than by
# scrolling. Reads A10's JSON artifact from disk -- never imports the audit package, so this
# notebook doesn't grow a dependency on its internals (B7 emits the artifact for exactly this).
# A10 groups by *canonical* claim_type + alias-resolved subject; this notebook's own `groups`
# dict is raw (lowered subject, raw claim_type) -- matching on lowered subject alone is enough
# to carry the subject's rank over onto every raw group it owns.
if GROUP_INVENTORY_PATH.exists():
    inventory = json.loads(GROUP_INVENTORY_PATH.read_text(encoding="utf-8"))["groups"]
    rank_by_subject = {row["subject"].strip().lower(): row["subjectRank"] for row in inventory}
    UNRANKED = float("inf")

    def _sort_key(item):
        (subject, claim_type), idxs = item
        return (rank_by_subject.get(subject, UNRANKED), subject, claim_type)

    print(f"(ranked by A8 subject prominence, from {GROUP_INVENTORY_PATH.name})\n")
    for (subject, claim_type), idxs in sorted(groups.items(), key=_sort_key):
        distinct_sources = len({candidates[i]["source_id"] for i in idxs})
        rank = rank_by_subject.get(subject, UNRANKED)
        rank_label = f"rank {rank:>4}" if rank != UNRANKED else "rank  n/a"
        print(f"  {rank_label}  {subject!r:30} {claim_type!r:20} {len(idxs)} rows, {distinct_sources} distinct sources")
else:
    print(
        f"(no A10 inventory found at {GROUP_INVENTORY_PATH} -- run "
        "`python -m audit.group_inventory --db` first for a prominence-ranked listing; "
        "falling back to alphabetical)\n"
    )
    for (subject, claim_type), idxs in sorted(groups.items()):
        distinct_sources = len({candidates[i]["source_id"] for i in idxs})
        print(f"  {subject!r:30} {claim_type!r:20} {len(idxs)} rows, {distinct_sources} distinct sources")

In [ ]:
from extraction.run_extraction import _claim_key, REJECTED_TRUST_TIER


def review_group(subject: str, claim_type: str) -> None:
    """Prints every candidate row for one (subject, claim_type) group, with its
    source segment text. Call this repeatedly for whichever groups you're reviewing
    right now -- output stays small per call, so no IOPub flooding. Ends with two
    ready-to-paste lines covering every row shown: `approved_keys += [...]` and
    `rejected_keys += [...]` -- paste whichever below, delete the rows that don't
    belong in that list, then run.

    Track C2 (DEV-104): keyed by the same 5-tuple identity DEV-101's merge-on-write
    uses (`extraction.run_extraction._claim_key` -- imported, never redefined), not
    by position. `variant_claims_candidates.json` is rewritten in extraction order
    on every re-extraction, so an index recorded in one review session can point at
    an entirely different, unrelated claim by the next session -- proven live, not
    hypothetical (see the markdown cell above the promotion cells below).

    Track F3 (DEV-113): a row's `trust_tier` already tells you whether a previous
    session reviewed it -- 1 means promoted, 2 means checked-and-rejected (an
    extraction error caught against the source text), 3 means never reviewed. Rows
    already at 1 or 2 are marked inline so you don't re-litigate a decision someone
    already made; they're still listed (not hidden) so the group reads complete."""
    key = (subject.strip().lower(), claim_type)
    idxs = groups.get(key)
    if not idxs:
        print(f"no group for {subject!r}/{claim_type!r} -- check the listing above for the exact claim_type spelling")
        return
    row_keys = []
    for i in idxs:
        c = candidates[i]
        row_key = _claim_key(c)
        row_keys.append(row_key)
        tier = c["trust_tier"]
        status = {1: " [ALREADY PROMOTED]", REJECTED_TRUST_TIER: " [ALREADY REJECTED]"}.get(tier, "")
        print("=" * 80)
        print(f"[{i}] key={row_key}{status}")
        print(f"    subject={c['subject_name']!r}  claim_type={c['claim_type']!r}")
        print(f"    claim_value: {c['claim_value']}")
        print(f"    source_id={c['source_id']}  passage_ref={c['passage_ref']}  trust_tier={tier}")
        segment_text = segment_maps.get(c["source_id"], {}).get(c["passage_ref"])
        if segment_text is None:
            print("    [!] no matching segment for this passage_ref -- verify manually against the source text")
        else:
            # preview = segment_text[:400] + ("..." if len(segment_text) > 400 else "")
            preview = segment_text
            print(f"    segment text: {preview}")
    print("-" * 80)
    print("copy/paste (delete rows that don't belong in each list):")
    print(f"approved_keys += {row_keys!r}")
    print(f"rejected_keys += {row_keys!r}")


# Example -- start with the floor conflicts:
review_group("Io", "parentage")

## Track B5 (Stage P5) — `review_passage`: whole-passage review

`review_group` above reviews one `(subject, claim_type)` group at a time, re-printing
the same segment text every time two groups share a passage. Stage P5's backlog is
1,059 distinct `(source_id, passage_ref)` pairs holding all 6,695 tier-3 rows —
passage-scoped review reads each segment once and adjudicates every row cited to it
in that one pass (Context, "the backlog is 1,059 passage reads").

`review_passage(source_id, passage_ref)` is the passage-scoped counterpart: it prints
the segment **once**, then every tier-3 row cited to it, each annotated by its B3
bucket (`ingestion/extraction/claim_evidence.py`) —

- **A** — the kinship formula is attested verbatim in this segment; the printed
  `matched span` is displayed evidence, batch-approvable per ADR-004 Amendment 1 (B7).
- **C** — both the subject and the claimed parent are named in the segment, but no
  formula matched; read required.
- **D** — only one of the two is named; read required until B2a reports the alias
  layers clean (an "absent" party might just be a spelling gap).
- **E** — neither is named; rejection-leaning, but still needs an opened segment to
  batch-reject except through the separate, budgeted C5 sweep (B7's carve-out).
- **Z** (`Z_JUNK` / `Z_BLOCKED` / `Z_HOLD`) — classified before any read; the subject
  itself cannot seed (placeholder, D4 namesake exclusion, or not yet an entity).

Ends with the same pasteable `approved_keys += [...]` / `rejected_keys += [...]`
lines `review_group` does, keyed by the identical `_claim_key` 5-tuple — the write
cell further down needs no changes to serve either function. `passage_queue` below is
`build_passage_queue` (B4): passages sorted by A6-contested row count first, then
total row count, so the read cost lands on conflicts before anything else.

In [ ]:
import psycopg2

import config
from extraction.claim_evidence import Bucket, build_alias_maps, build_passage_queue, bucket_claim, load_contested_keys

_conn = psycopg2.connect(
    host=config.POSTGRES_HOST, port=config.POSTGRES_PORT, dbname=config.POSTGRES_DB,
    user=config.POSTGRES_USER, password=config.POSTGRES_PASSWORD,
)
with _conn.cursor() as _cur:
    _cur.execute("SELECT name FROM entities")
    known_names = {row[0] for row in _cur.fetchall()}
name_aliases, spelling_aliases = build_alias_maps(_conn)
contested_keys = load_contested_keys(None, _conn)
_conn.close()

passage_queue = build_passage_queue(candidates, contested_keys)
print(f"{len(passage_queue)} passage(s) queued, {sum(e.total_rows for e in passage_queue)} tier-3 row(s) total")
print("top 10 by contested-row count:")
for entry in passage_queue[:10]:
    print(f"  {entry.source_id:28} {entry.passage_ref:16} contested={entry.contested_count:3} total={entry.total_rows:3}")

In [ ]:
_BUCKET_LABEL = {
    Bucket.Z_JUNK: "Z (junk subject -- reject mechanically at trust_tier=2)",
    Bucket.Z_BLOCKED: "Z (D4 namesake exclusion -- enters the bucket-Z blocked register)",
    Bucket.Z_HOLD: "Z (subject not yet an entity -- hold for Track D, re-queue once it exists)",
    Bucket.A: "A (kinship formula attested verbatim -- batch-approvable, ADR-004 Amendment 1)",
    Bucket.C: "C (both parties present, no formula matched -- read required)",
    Bucket.D: "D (one party absent -- read required until B2a reports the alias layers clean)",
    Bucket.E: "E (neither party present -- rejection-leaning, needs an opened segment or C5)",
    Bucket.UNPARSED: "UNPARSED (claim value has no resolvable <relation> formula -- read)",
}


def review_passage(source_id: str, passage_ref: str) -> None:
    """Prints the cited segment ONCE, then every tier-3 row cited to it, each
    annotated by its `claim_evidence.bucket_claim` bucket. Ends with two
    ready-to-paste lines covering every row shown -- `approved_keys += [...]` and
    `rejected_keys += [...]` -- paste whichever below, delete the rows that don't
    belong in that list, then run (same `approved_keys`/`rejected_keys`/write cell
    as `review_group` -- one read, N adjudications, per this stage's B5 exit
    criterion)."""
    rows = [
        c
        for c in candidates
        if c["source_id"] == source_id and c["passage_ref"] == passage_ref and c.get("trust_tier", 3) == 3
    ]
    if not rows:
        print(f"no tier-3 rows cited to {source_id} {passage_ref!r}")
        return

    segment_text = segment_maps.get(source_id, {}).get(passage_ref)
    print("=" * 80)
    print(f"{source_id} {passage_ref} -- {len(rows)} tier-3 row(s)")
    print("=" * 80)
    if segment_text is None:
        print("[!] no matching segment for this passage_ref -- verify manually against the source text")
    else:
        print(segment_text)
    print("=" * 80)

    row_keys = []
    for c in rows:
        row_key = _claim_key(c)
        row_keys.append(row_key)
        b = bucket_claim(c, segment_text or "", known_names, name_aliases, spelling_aliases)
        print(f"[{_BUCKET_LABEL[b.bucket]}]")
        print(f"    key={row_key}")
        print(f"    subject={c['subject_name']!r}  claim_type={c['claim_type']!r}  claim_value={c['claim_value']!r}")
        if b.evidence_span:
            print(f"    matched span: {b.evidence_span!r}")
        print("-" * 80)

    print("copy/paste (delete rows that don't belong in each list):")
    print(f"approved_keys += {row_keys!r}")
    print(f"rejected_keys += {row_keys!r}")


# Example -- the top-contested passage on today's queue:
if passage_queue:
    review_passage(passage_queue[0].source_id, passage_queue[0].passage_ref)

## Decide which rows to promote or reject

Edit the list below with the keys (from the `approved_keys += [...]` line printed by
`review_group`) you approve this session -- run it, then repeat across as many sessions as you
like. **Additive only**: this only ever sets rows to `trust_tier=1`, never demotes an
already-promoted row back down, so partial review across multiple sittings is safe -- you don't
need to re-list keys you approved in an earlier session.

**Track C (DEV-104): keyed, not positional.** Promotion here used to be `candidates[i]["trust_tier"]
= 1` against `approved_indices: list[int]` -- unsafe the moment this file is ever regenerated by a
re-extraction, since `variant_claims_candidates.json` is rewritten in extraction order (DEV-101),
so index `i` can silently point at a different claim next time. The cell below proves this
happened, not just that it could.

**Track F3 (DEV-113): rejections get the same treatment, at `trust_tier=2`.** A row you check
against its source and find unsupported (wrong subject, reversed direction, misread Homeric
formula, ...) belongs in `rejected_keys` below, not just deletion from `approved_keys` -- an
unlisted row and a rejected row look identical to `review_group` next session (`trust_tier=3`,
"unreviewed") unless the rejection is written down. `seedgen` already ignores anything that
isn't exactly `trust_tier==1`, and `run_extraction.py`'s merge-on-write already preserves any
non-default tier across a re-extraction (`REJECTED_TRUST_TIER=2 != DEFAULT_TRUST_TIER=3`) -- both
for free, so no seedgen/extraction code changes were needed for this to work.

In [ ]:
# <-- EDIT THIS after reviewing the rows above
# Keyed by (subject_name, claim_type, claim_value, source_id, passage_ref) -- the same 5-tuple
# DEV-101's merge-on-write uses (extraction.run_extraction._CLAIM_IDENTITY) -- never by position.
approved_keys: list[tuple] = []

# Track F3 (DEV-113): rows checked against their source and found unsupported go here instead,
# at trust_tier=2 -- so the next reviewer sees "checked, rejected" rather than "not yet reviewed".
rejected_keys: list[tuple] = []

### Historical promotions, migrated from position to key (Track C1/C3, DEV-104)

**C1 -- the hazard, proven, not assumed.** The three `approved_indices += [...]` cells below used
to record this notebook's early review sessions by position. Replaying `run_extraction.py`'s real
merge-on-write (`_write_claims_preserving_review`) against a reordered copy of the live 7,429-row
file -- same claims, different order, exactly what a re-extraction with a different segment
processing order produces -- showed **73 of 73** of those historical indices now pointing at a
completely different, unrelated claim (e.g. index `743` used to be `('Aphrodite', 'parentage',
'child of Zeus', ...)`; after the reorder it resolves to `('Agamemnon', 'death', 'killed by
Aegisthus', ...)`). No error, no warning -- exactly ADR-004's gate failing silently. DEV-101's
key-based merge itself worked correctly throughout (all 71 promoted **keys** survived the
reorder); the hazard is specific to a notebook that re-derives *positions* across sessions.

**C3 -- the migration found a real drift, not just a hypothetical one.** Resolving the three
historical cells' indices against **today's** live file (no simulation) already disagrees with
what's actually promoted: only **71 of those 73** indices currently resolve to a `trust_tier=1`
row, and one promoted row (`Aphrodite`/`birth`) isn't covered by any of the three cells at all. Per
this track's own instruction ("a mismatch means the file already drifted -- stop and triage, do
not proceed"): the three stale index cells below are **not** used as the migration source. The
cell below instead lists the **71 keys the live file actually has at `trust_tier=1` right now** --
the real, current review decision, which DEV-101's merge has been correctly preserving by key
since it landed -- across exactly the 4 groups the Contracts section names: `Aphrodite/parentage`
(33), `Aphrodite/birth` (1), `Achilles/death` (27), `Io/parentage` (10).

In [ ]:
approved_keys += [
    ('Achilles', 'death', 'Achilles was slain; Ajax carried his dead body from the battlefield', 'ovid-metamorphoses', '13.280-13.312'),
    ('Achilles', 'death', 'Achilles wished Hector had slain him instead of dying swept away by the river', 'homer-iliad', '21.272-21.297'),
    ('Achilles', 'death', 'Foretold by dying Hector to be slain by Paris and Phoebus Apollo at the Scaean gate', 'homer-iliad', '22.328-22.374'),
    ('Achilles', 'death', 'Killed by Paris and Apollo with arrows/shafts', 'ovid-metamorphoses', '13.481-13.507'),
    ('Achilles', 'death', "Killed by an arrow shot by Paris, guided by Apollo, at Apollo's and Neptune's instigation - a divergence from the more common tradition of Achilles being wounded in the heel", 'ovid-metamorphoses', '12.580-12.612'),
    ('Achilles', 'death', "Thetis foretells his death will follow soon after Hector's death", 'homer-iliad', '18.50-18.96'),
    ('Achilles', 'death', 'Thetis foretold he would perish beneath the wall of the Trojans by the swift missiles of Apollo', 'homer-iliad', '21.272-21.297'),
    ('Achilles', 'death', 'Thetis forewarned Achilles that he would die by the hand of Apollo if he slew Tenes', 'apollodorus-bibliotheca', 'E.3.20-E.3.26'),
    ('Achilles', 'death', 'denied by unjust fate to both the Greeks and Ulysses, implying his death before the contest for his arms', 'ovid-metamorphoses', '13.128-13.161'),
    ('Achilles', 'death', 'destined to meet his death and fate at some future point', 'homer-iliad', '20.292-20.339'),
    ('Achilles', 'death', 'destined to perish in deep-soiled Troy, far from his native land', 'homer-iliad', '24.77-24.119'),
    ('Achilles', 'death', 'doomed to a speedy death, brief span of life', 'homer-iliad', '1.386-1.427'),
    ('Achilles', 'death', 'doomed to be brought low beneath the wall of the wealthy Trojans', 'homer-iliad', '23.54-23.98'),
    ('Achilles', 'death', 'fated to be slain in fight by a god and a mortal, far from his father and mother', 'homer-iliad', '19.392-20.11'),
    ('Achilles', 'death', 'foretold that he will die in battle, smitten by a spear cast or an arrow from the string', 'homer-iliad', '21.97-21.135'),
    ('Achilles', 'death', 'killed by Alexander', 'apollodorus-bibliotheca', 'E.4.6-E.5.4'),
    ('Achilles', 'death', 'killed by Apollo', 'apollodorus-bibliotheca', 'E.3.20-E.3.26'),
    ('Achilles', 'death', 'killed by Apollo', 'apollodorus-bibliotheca', 'E.4.6-E.5.4'),
    ('Achilles', 'death', 'killed by Apollo', 'homer-iliad', '19.392-20.11'),
    ('Achilles', 'death', 'killed by Apollo', 'homer-iliad', '21.272-21.297'),
    ('Achilles', 'death', 'killed by Apollo', 'homer-iliad', '22.328-22.374'),
    ('Achilles', 'death', 'killed by Apollo', 'ovid-metamorphoses', '13.481-13.507'),
    ('Achilles', 'death', 'killed by Paris', 'ovid-metamorphoses', '12.580-12.612'),
    ('Achilles', 'death', 'killed by Paris', 'ovid-metamorphoses', '13.162-13.204'),
    ('Achilles', 'death', 'killed by Paris', 'ovid-metamorphoses', '13.481-13.507'),
    ('Achilles', 'death', 'slain at Troy', 'homer-odyssey', '3.102-3.154'),
    ('Achilles', 'death', 'slain in the land of Troy, far from Argos', 'homer-odyssey', '24.35-24.84'),
    ('Aphrodite', 'birth', 'Born from the sea foam formed by the severed members of Heaven, cast into the sea by Cronus', 'hesiod-theogony', '176-232'),
    ('Aphrodite', 'parentage', 'child of Dione', 'apollodorus-bibliotheca', '1.3.1-1.3.5'),
    ('Aphrodite', 'parentage', 'child of Dione', 'homer-iliad', '5.334-5.380'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'apollodorus-bibliotheca', '1.3.1-1.3.5'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'hesiod-homeric-hymns', '3.182-3.238'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'hesiod-homeric-hymns', '5.106-5.154'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'hesiod-homeric-hymns', '5.53-5.105'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '14.153-14.196'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '14.197-14.241'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '20.103-20.143'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '3.340-3.380'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '5.297-5.333'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '5.334-5.380'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-iliad', '5.778-5.824'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-odyssey', '8.256-8.294'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'homer-odyssey', '8.295-8.342'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'ovid-metamorphoses', '14.566-14.608'),
    ('Aphrodite', 'parentage', 'child of Zeus', 'ovid-metamorphoses', '15.807-15.842'),
    ('Aphrodite', 'parentage', 'daughter of Dione', 'homer-iliad', '5.334-5.380'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'hesiod-homeric-hymns', '3.182-3.238'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'hesiod-homeric-hymns', '5.106-5.154'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'hesiod-homeric-hymns', '5.155-5.201'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'hesiod-homeric-hymns', '5.53-5.105'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '14.153-14.196'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '14.197-14.241'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '21.382-21.422'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '23.161-23.191'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '3.340-3.380'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '5.121-5.165'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '5.297-5.333'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '5.334-5.380'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-iliad', '5.778-5.824'),
    ('Aphrodite', 'parentage', 'daughter of Zeus', 'homer-odyssey', '8.295-8.342'),
    ('Aphrodite', 'parentage', 'daughter of Zeus (the mighty son of Cronos)', 'homer-odyssey', '8.256-8.294'),
    ('Io', 'parentage', 'Io is the daughter of Inachus', 'ovid-metamorphoses', '1.622-1.667'),
    ('Io', 'parentage', 'child of Iasus', 'apollodorus-bibliotheca', '2.1.2-2.1.3'),
    ('Io', 'parentage', 'child of Inachus', 'apollodorus-bibliotheca', '2.1.2-2.1.3'),
    ('Io', 'parentage', 'child of Inachus', 'ovid-metamorphoses', '1.567-1.600'),
    ('Io', 'parentage', 'child of Inachus', 'ovid-metamorphoses', '1.622-1.667'),
    ('Io', 'parentage', 'child of Piren', 'apollodorus-bibliotheca', '2.1.2-2.1.3'),
    ('Io', 'parentage', 'daughter of Iasus, according to the main narrative', 'apollodorus-bibliotheca', '2.1.2-2.1.3'),
    ('Io', 'parentage', 'daughter of Inachus', 'ovid-metamorphoses', '1.567-1.600'),
    ('Io', 'parentage', 'daughter of Inachus, according to Castor and many tragedians', 'apollodorus-bibliotheca', '2.1.2-2.1.3'),
    ('Io', 'parentage', 'daughter of Piren, according to Hesiod and Acusilaus', 'apollodorus-bibliotheca', '2.1.2-2.1.3'),
]

### Stage P4 Track F1 — Batch 1 (DEV-110): 95 rows promoted across 26 groups

Tranche picked by F0a's Tier-1 rule (top-20 prominence subject + a canonical claim_type with zero
promoted coverage): every `marriage` and `epithet` group for the 20 subjects frozen in F0a
(`Odysseus, Zeus, Achilles, Heracles, Hector, Agamemnon, Apollo, Telemachus, Diomedes, Patroclus,
Athena, Menelaus, Penelope, Poseidon, Hermes, Aphrodite, Priam, Hera, Nestor, Ajax` — `Telemachus`/
`Patroclus`/`Ajax` turned out to have no `marriage`/`epithet` candidates at all). 26 groups, 17
distinct top-20 subjects covered, both `marriage` and `epithet` promoted to canonical claim_type #3
and #4 (`parentage`, `death` already covered) — the "≥4 canonical claim_types" exit gate is met by
this batch alone.

Every row was checked against its actual source snippet before promotion, not promoted by
`claim_value` string alone — this caught **5 real extraction errors**, none promoted:
- `Achilles married_to Peleus` (`homer-iliad` 17.412-17.455) — the passage is about the immortal
  horses given to **Peleus** (Achilles' father) at *his* wedding to Thetis, not a claim about
  Achilles' own marriage at all.
- `Aphrodite married_to Arges` (`hesiod-theogony` 924-955) — the cited passage is Athena's birth
  from Zeus's head; does not mention Aphrodite marrying anyone.
- `Poseidon married_to Medus` (`hesiod-theogony` 270-305) — the passage names **Medusa**, not
  "Medus"; likely a name-truncation extraction bug.
- `Menelaus married_to Hellen` — a spelling collision with the *other* Hellen (son of Deucalion,
  ancestor of the Hellenes); the same group has several correctly-spelled `Helen` rows from the
  same passages, promoted instead.
- `Zeus married_to Herse` (`homer-odyssey` 15.92-15.142) — the passage is Telemachus invoking "Zeus,
  the loud-thundering lord of Here [Hera]"; Herse does not appear in it at all.

Breadth over depth per F0a: near-duplicate rephrasings of the same fact (e.g. a dozen ways of
saying "wife of Odysseus") were not all promoted — one or two per distinct fact per source is
enough for citation diversity without spending review effort for zero additional coverage (the
*Contracts* section's own dedup-waste finding). Genuinely distinct claims (different named spouse,
different epithet) were kept in full — several groups (`Zeus`/marriage, `Odysseus`/marriage,
`Priam`/marriage) surface real cross-tradition variety this way. Applied via the script-equivalent
of the cell below's own C2/C4/C5 logic (95 keys, all resolved, zero unmatched); the promoted keys
are recorded in `ingestion/audit/promotion_log.json` under `batchLabel: "p4-f1-batch1"`.

In [ ]:
approved_keys += [
    ('Achilles', 'epithet', 'swift-footed', 'homer-iliad', '1.364-1.385'),
    ('Achilles', 'epithet', 'lordly son of Peleus', 'homer-iliad', '10.349-10.399'),
    ('Achilles', 'epithet', 'son of Aeacus', 'homer-iliad', '10.400-10.445'),
    ('Achilles', 'epithet', 'son of Peleus', 'homer-odyssey', '5.291-5.332'),
    ('Achilles', 'marriage', 'married to Medea', 'apollodorus-bibliotheca', 'E.5.5-E.5.13'),
    ('Agamemnon', 'marriage', 'married to Clytaemnestra', 'apollodorus-bibliotheca', 'E.2.15-E.3.5'),
    ('Agamemnon', 'marriage', 'wedded wife Clytemnestra', 'homer-iliad', '1.101-1.146'),
    ('Agamemnon', 'marriage', 'married to Cassandra', 'apollodorus-bibliotheca', 'E.5.22-E.6.4'),
    ('Aphrodite', 'epithet', 'Called Aphrodite, foam-born goddess, rich-crowned Cytherea, Cyprogenes, and Philommedes', 'hesiod-theogony', '176-232'),
    ('Aphrodite', 'epithet', 'called the Cyprian and rich-crowned Cytherea', 'hesiod-homeric-hymns', '4.550-5.6'),
    ('Aphrodite', 'epithet', 'Ruler of the walled cities of sea-set Cyprus, called Cytherea, gold-crowned and beautiful', 'hesiod-homeric-hymns', '6.1-7.31'),
    ('Aphrodite', 'epithet', 'the Idalian, who hates cruel hearts', 'ovid-metamorphoses', '14.675-14.715'),
    ('Aphrodite', 'marriage', 'married to Hephaestus', 'homer-odyssey', '8.256-8.294'),
    ('Aphrodite', 'marriage', 'married to Anchises', 'hesiod-homeric-hymns', '5.106-5.154'),
    ('Aphrodite', 'marriage', 'married to Adonis', 'ovid-metamorphoses', '10.503-10.542'),
    ('Apollo', 'epithet', 'shoots afar; god before whom the gods tremble and rise', 'hesiod-homeric-hymns', '2.483-3.29'),
    ('Apollo', 'epithet', 'Phoebus', 'hesiod-homeric-hymns', '4.68-4.114'),
    ('Apollo', 'epithet', 'Far-shooter, Phoebus', 'hesiod-homeric-hymns', '4.387-4.432'),
    ('Apollo', 'epithet', 'the god that smiteth afar', 'homer-iliad', '7.54-7.102'),
    ('Apollo', 'epithet', 'Phoebus Apollo of the golden sword', 'homer-iliad', '15.246-15.280'),
    ('Apollo', 'epithet', 'Apollo Delphinius, named for his dolphin form', 'hesiod-homeric-hymns', '3.474-3.523'),
    ('Athena', 'epithet', 'guardian of the city, loves deeds of war with Ares', 'hesiod-homeric-hymns', '9.1-16.1'),
    ('Athena', 'epithet', 'Tritogeneia', 'hesiod-homeric-hymns', '28.1-30.1'),
    ('Athena', 'epithet', 'Alalcomenean Athene', 'homer-iliad', '4.1-4.49'),
    ('Diomedes', 'marriage', 'married to Aegialia', 'apollodorus-bibliotheca', '1.8.6-1.9.1'),
    ('Diomedes', 'marriage', 'married to Aegialeia, daughter of Adrastus', 'homer-iliad', '5.381-5.425'),
    ('Diomedes', 'marriage', "married to Daunus' daughter", 'ovid-metamorphoses', '14.483-14.526'),
    ('Hector', 'marriage', 'married to Andromache', 'apollodorus-bibliotheca', '3.12.6'),
    ('Hector', 'marriage', 'married to Andromache', 'homer-iliad', '6.466-6.502'),
    ('Hera', 'epithet', 'Argive Hera', 'homer-iliad', '4.1-4.49'),
    ('Hera', 'epithet', 'Queen of Gods, glorious child of Saturn', 'ovid-metamorphoses', '2.508-2.541'),
    ('Hera', 'marriage', 'married to Zeus', 'hesiod-homeric-hymns', '9.1-16.1'),
    ('Hera', 'marriage', 'married to Zeus', 'homer-iliad', '14.153-14.196'),
    ('Hera', 'marriage', "married to Zeus, first came together in love without their parents' knowledge", 'homer-iliad', '14.277-14.311'),
    ('Hera', 'marriage', 'Juno married Jupiter (Jove), her brother', 'ovid-metamorphoses', '9.487-9.516'),
    ('Heracles', 'marriage', 'married to Megaera', 'apollodorus-bibliotheca', '2.4.11-2.4.12'),
    ('Heracles', 'marriage', 'married to Deianira', 'apollodorus-bibliotheca', '2.7.4-2.7.6'),
    ('Heracles', 'marriage', 'married to Deianira', 'ovid-metamorphoses', '9.172-9.210'),
    ('Heracles', 'marriage', 'married to Hebe', 'apollodorus-bibliotheca', '2.7.7'),
    ('Heracles', 'marriage', 'married to Hebe', 'homer-odyssey', '11.582-11.626'),
    ('Heracles', 'marriage', 'gave Megara to Iolaus and sought to wed Iole', 'apollodorus-bibliotheca', '2.5.12-2.6.1'),
    ('Hermes', 'epithet', 'Slayer of Argus', 'hesiod-homeric-hymns', '4.68-4.114'),
    ('Hermes', 'epithet', 'called Argeiphontes', 'homer-iliad', '24.77-24.119'),
    ('Hermes', 'epithet', 'Argeiphontes', 'homer-odyssey', '10.274-10.320'),
    ('Hermes', 'marriage', 'married to Daughter of Dryops', 'hesiod-homeric-hymns', '19.27-23.1'),
    ('Hermes', 'marriage', 'married to Polymele', 'homer-iliad', '16.155-16.197'),
    ('Menelaus', 'epithet', 'dear to Ares', 'homer-iliad', '3.38-3.83'),
    ('Menelaus', 'epithet', 'fostered of Zeus', 'homer-iliad', '10.42-10.85'),
    ('Menelaus', 'epithet', 'son of Atreus', 'homer-iliad', '23.429-23.472'),
    ('Menelaus', 'marriage', 'married to Helen, contested by Paris who took her', 'homer-iliad', '3.245-3.291'),
    ('Menelaus', 'marriage', 'married to Helen', 'homer-odyssey', '4.265-4.314'),
    ('Menelaus', 'marriage', 'has Helen to wife, husband of the daughter of Zeus', 'homer-odyssey', '4.538-4.586'),
    ('Nestor', 'epithet', 'king of sandy Pylos', 'homer-iliad', '2.48-2.83'),
    ('Nestor', 'epithet', 'tamer of horses', 'homer-odyssey', '2.405-3.20'),
    ('Nestor', 'marriage', 'married to Anaxibia daughter of Cratieus', 'apollodorus-bibliotheca', '1.9.8-1.9.11'),
    ('Nestor', 'marriage', 'married to Eurydice', 'homer-odyssey', '3.447-3.477'),
    ('Odysseus', 'epithet', 'Odysseus of many wiles', 'homer-iliad', '14.64-14.108'),
    ('Odysseus', 'epithet', 'Zeus-born', 'homer-odyssey', '2.309-2.360'),
    ('Odysseus', 'epithet', 'fostered of Zeus', 'homer-odyssey', '10.229-10.273'),
    ('Odysseus', 'marriage', 'married to Penelope', 'apollodorus-bibliotheca', 'E.3.6-E.3.12'),
    ('Odysseus', 'marriage', 'husband of Penelope, who longs for his return', 'homer-odyssey', '2.224-2.266'),
    ('Odysseus', 'marriage', 'married to Hecuba', 'apollodorus-bibliotheca', 'E.5.22-E.6.4'),
    ('Odysseus', 'marriage', 'married to Circe', 'apollodorus-bibliotheca', 'E.7.9-E.7.16'),
    ('Odysseus', 'marriage', 'married to Callidice', 'apollodorus-bibliotheca', 'E.7.33-E.7.40'),
    ('Odysseus', 'marriage', "married to Thoas' daughter", 'apollodorus-bibliotheca', 'E.7.33-E.7.40'),
    ('Odysseus', 'marriage', 'married to Penelope, wife who wove and unraveled the shroud of Laertes to delay remarriage', 'homer-odyssey', '24.120-24.169'),
    ('Penelope', 'marriage', 'married to Odysseus', 'apollodorus-bibliotheca', '3.10.4-3.10.7'),
    ('Penelope', 'marriage', 'married to Odysseus', 'homer-odyssey', '16.68-16.111'),
    ('Poseidon', 'epithet', 'loud-roaring Holder of the Earth, whose sacred grove is at Onchestus', 'hesiod-homeric-hymns', '4.162-4.211'),
    ('Poseidon', 'epithet', 'Shaker of Earth', 'homer-iliad', '8.425-8.468'),
    ('Poseidon', 'epithet', 'dark-haired Earth-shaker, receiving sacrifice of black bulls at Pylos', 'homer-odyssey', '2.405-3.20'),
    ('Poseidon', 'marriage', 'married to Amphitrite', 'apollodorus-bibliotheca', '1.4.3-1.5.2'),
    ('Poseidon', 'marriage', 'married to Amymone', 'apollodorus-bibliotheca', '2.1.4'),
    ('Poseidon', 'marriage', 'married to Thoosa', 'homer-odyssey', '1.44-1.95'),
    ('Priam', 'marriage', 'married to Arisbe', 'apollodorus-bibliotheca', '3.12.5'),
    ('Priam', 'marriage', 'married to Hecuba', 'apollodorus-bibliotheca', '3.12.5'),
    ('Priam', 'marriage', 'married to Hecuba', 'homer-iliad', '24.200-24.246'),
    ('Priam', 'marriage', 'married to Castianeira', 'homer-iliad', '8.292-8.334'),
    ('Priam', 'marriage', 'married to Laothoe', 'homer-iliad', '22.25-22.58'),
    ('Zeus', 'epithet', 'dark-clouded Son of Cronos, far-seeing, loud-thunderer', 'hesiod-homeric-hymns', '2.434-2.482'),
    ('Zeus', 'epithet', 'son of Cronos', 'homer-iliad', '4.240-4.271'),
    ('Zeus', 'epithet', 'father of men and gods', 'homer-iliad', '8.41-8.77'),
    ('Zeus', 'epithet', 'who hurls the thunderbolt', 'homer-odyssey', '14.235-14.284'),
    ('Zeus', 'epithet', "the stranger's god", 'homer-odyssey', '14.235-14.284'),
    ('Zeus', 'marriage', 'married to Hera', 'apollodorus-bibliotheca', '1.3.1-1.3.5'),
    ('Zeus', 'marriage', 'married to Hera', 'homer-iliad', '14.312-14.360'),
    ('Zeus', 'marriage', 'married to Metis', 'apollodorus-bibliotheca', '1.3.6-1.4.2'),
    ('Zeus', 'marriage', 'married to Alcmena', 'apollodorus-bibliotheca', '2.4.7-2.4.8'),
    ('Zeus', 'marriage', 'married to Mnemosyne', 'hesiod-theogony', '53-103'),
    ('Zeus', 'marriage', 'married to Themis', 'hesiod-theogony', '869-923'),
    ('Zeus', 'marriage', 'married to Eurynome', 'hesiod-theogony', '869-923'),
    ('Zeus', 'marriage', 'married to Demeter', 'hesiod-homeric-hymns', '2.19-2.58'),
    ('Zeus', 'marriage', 'married to Leto', 'hesiod-theogony', '869-923'),
    ('Zeus', 'marriage', 'married to Maia', 'hesiod-homeric-hymns', '4.28-4.67'),
    ('Zeus', 'marriage', 'married to Leda', 'hesiod-homeric-hymns', '33.1'),
]

### Stage P4 Track F2 — Batch 2 (DEV-112): 107 rows promoted across 25 canonical groups, closes the top-20-subject-coverage gate to 20/20

Tranche picked by F0a's rule over the *refreshed* A8/A10 output (re-run first, per F2's own
template instruction): with `marriage`/`epithet` already covering 4/4 canonical claim_types
after F1, Tier 1 for this batch is every top-20 subject's `notable_claim` group (plus a few
smaller `other`/`burial`/`role` groups) — 16 subjects had one, `Telemachus`/`Patroclus`/`Ajax`
did not. `Ajax` turned out to have a `notable_claim` group after all (missed in F1's tier scan
because its rows are filed under `subject_name='Ajax'` while other Ajax-referencing rows in the
candidate file use `'Ajax son of Telamon'` — a name-variant split worth a future J-item, not
fixed here). `Telemachus` and `Patroclus` have **no** unpromoted-type candidates at all, so they
were covered via Tier 3 instead — their `parentage`/`death` groups (already-canonical types),
exactly where F0a's rule predicted GAP-001's backlog would surface.

Same per-row verification discipline as F1, and it kept paying off: **8 more extraction errors
caught and not promoted**, a new failure shape this batch — a Homeric formula ("Zeus, son of
Cronos" as an address to heralds/companions) misread as a literal parentage claim, for **both**
`Telemachus` (`'child of Zeus'`, `homer-odyssey` 21.96-21.142) and `Patroclus` (`'child of
Zeus'`/`'sprung from Zeus'`, `homer-iliad` 1.325-1.363) — Achilles calling the heralds "you who
are guilty in my sight" and addressing them via a formulaic epithet, not a genealogy statement.
`Telemachus`/`'child of Athena'` (`homer-odyssey` 2.267-2.308) is the same shape: Athena
approaching in the guise of his mentor Mentor, not a birth claim. `Patroclus`/`'killed by
Sarpedon'` (`homer-iliad` 23.797-23.835) is **reversed** — the cited text shows Patroclus
stripping *Sarpedon's* armor, i.e. Patroclus killed Sarpedon, not the other way around. Three
more `Diomedes`/`notable_claim` rows were skipped as unverified (a probable `Diomede`/`Diomedes`
name confusion for one, two others whose cited passage showed a different speaker/scene than the
claim described), and one `Aphrodite`/`notable_claim` row (`'went to Olympus to ask Zeus for the
maidens' marriage'`) was skipped as a likely Athena-not-Aphrodite misattribution — the cited
Odyssey passage (20.54-20.101) is Odysseus's pre-slaughter prayer scene, where Athena is the
usual divine visitor, not Aphrodite.

Full promoted-key list: `ingestion/audit/promotion_log.json`, `batchLabel: "p4-f2-batch2"` (107
keys, 41 raw-`(subject_name, claim_type)` groups by the log's own un-normalized counting — 25
canonical `(resolved subject, normalized claim_type)` groups by F0a's tier-selection counting;
the difference is entirely subjects whose `notable_claim` rows span multiple raw spellings like
`notable`/`notable_act`/`notable_deed`, same shape DEV-107/Track G already collapsed for the
`claim_type_aliases` table). Not re-embedded as a literal cell here to avoid roughly doubling
this notebook's size for a second time in two batches — the promotion log and `DEVIATIONS.md`
#DEV-112 are the canonical records.

### Stage P4 Track F3 (DEV-113): `trust_tier=2` introduced; 31 rows from F1/F2 marked rejected

F1 and F2's own per-row review caught 13 distinct bad claims (5 in F1, 8 in F2, documented in the
markdown cells above and in `DEVIATIONS.md` #DEV-110/#DEV-112) but left every one of them at
`trust_tier=3` -- indistinguishable from "never reviewed" to the next session. `REJECTED_TRUST_TIER
= 2` (`extraction/run_extraction.py`) fixes that: a row a reviewer checked and found unsupported now
reads as "checked, rejected" instead of silently going back into the unreviewed pool.

Applied retroactively for all 14 already-caught patterns (31 underlying rows -- most patterns are
one row, but `Menelaus married_to Hellen` alone accounts for 18, one per passage where the *other*
Hellen collides on spelling with Helen of Troy): 13 patterns from F1/F2's own review-cell markdown
(5 marriage/parentage errors in F1, 8 formula-misread/reversed-direction/unverified errors in F2)
plus the third of F2's "three unverified `Diomedes`/`notable_claim` rows" that wasn't individually
named in the F2 cell's prose (`'exchanged arms with Glaucus after recall'`, `apollodorus-
bibliotheca` E.3.34-E.4.5). No new review was performed for this batch -- every row was already
checked against its source during F1/F2; this batch only records that existing decision in the
data. Full list in `ingestion/audit/promotion_log.json` under `batchLabel:
"p4-f3-retroactive-rejections"`.

Verified zero-impact on the seeded table: `seedgen`'s promotion filter is `trust_tier == 1` (exact
equality), so `trust_tier=2` rows were already excluded before this change (they were `3`) and stay
excluded now (`2`) -- `python -m seedgen --strict` produces a byte-identical `V12` (still 245 rows)
before and after. `run_extraction.py`'s merge-on-write (`_write_claims_preserving_review`, DEV-101)
also needed no change: its preserve check is `!= DEFAULT_TRUST_TIER` (not `== 1`), so it already
carries a `trust_tier=2` row across a re-extraction exactly like a promoted one.

Going forward, `review_group()` flags already-rejected rows inline (`[ALREADY REJECTED]`) and the
write cell below accepts a parallel `rejected_keys` list alongside `approved_keys` -- see the
updated "Decide which rows to promote or reject" section above.

In [ ]:
from datetime import datetime, timezone

from extraction.run_extraction import _claim_key, REJECTED_TRUST_TIER

PROMOTION_LOG_PATH = INGESTION_ROOT / "audit" / "promotion_log.json"
BATCH_LABEL = "p4-review-session"  # <-- EDIT before running: a short label for this session's batch

# Track C4 (DEV-104): a key that matches no row is an error, not a no-op -- refuse to write
# rather than silently skip it. The claim may have been dropped by a re-extraction (see the
# WARNING run_extraction.py prints when that happens) or simply mistyped.
key_to_index = {_claim_key(c): i for i, c in enumerate(candidates)}
unmatched = [k for k in approved_keys + rejected_keys if k not in key_to_index]
if unmatched:
    raise SystemExit(
        f"{len(unmatched)} key(s) match no row in {CANDIDATES_PATH.name} -- refusing to "
        "write:\n" + "\n".join(f"  {k}" for k in unmatched)
    )

# Track F3 (DEV-113): a key can't be both approved and rejected in the same session -- that's
# a contradiction, not a preference, so refuse rather than let one silently win.
both = set(approved_keys) & set(rejected_keys)
if both:
    raise SystemExit(f"{len(both)} key(s) listed in both approved_keys and rejected_keys:\n" + "\n".join(f"  {k}" for k in both))

newly_promoted = [k for k in approved_keys if candidates[key_to_index[k]]["trust_tier"] != 1]
newly_rejected = [k for k in rejected_keys if candidates[key_to_index[k]]["trust_tier"] != REJECTED_TRUST_TIER]
for k in approved_keys:
    candidates[key_to_index[k]]["trust_tier"] = 1
for k in rejected_keys:
    candidates[key_to_index[k]]["trust_tier"] = REJECTED_TRUST_TIER

CANDIDATES_PATH.write_text(json.dumps(candidates, indent=2, ensure_ascii=False), encoding="utf-8")
total_promoted = sum(1 for c in candidates if c["trust_tier"] == 1)
total_rejected = sum(1 for c in candidates if c["trust_tier"] == REJECTED_TRUST_TIER)
print(
    f"{len(newly_promoted)} row(s) newly promoted, {len(newly_rejected)} newly rejected this run; "
    f"{total_promoted} total at trust_tier=1, {total_rejected} total at trust_tier=2 "
    f"(out of {len(candidates)})"
)

# Track C5 (DEV-104): per-batch promotion log, appended (never overwritten/replaced) -- lets a
# batch's promotions be identified and reverted as a unit if Track F's compare.py comes back red.
# Only real, new decisions are logged -- re-running an already-decided key's line logs nothing,
# since nothing changed. Track F3 (DEV-113): rejections logged the same way, under "rejectedKeys",
# so a rejection batch is auditable/revertable exactly like a promotion batch.
if newly_promoted or newly_rejected:
    log_entries = json.loads(PROMOTION_LOG_PATH.read_text(encoding="utf-8")) if PROMOTION_LOG_PATH.exists() else []
    entry = {
        "batchLabel": BATCH_LABEL,
        "date": datetime.now(timezone.utc).isoformat(),
        "keys": [list(k) for k in newly_promoted],
        "groupCount": len({(k[0], k[1]) for k in newly_promoted}),
    }
    if newly_rejected:
        entry["rejectedKeys"] = [list(k) for k in newly_rejected]
        entry["rejectedGroupCount"] = len({(k[0], k[1]) for k in newly_rejected})
    log_entries.append(entry)
    PROMOTION_LOG_PATH.parent.mkdir(parents=True, exist_ok=True)
    PROMOTION_LOG_PATH.write_text(json.dumps(log_entries, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    print(
        f"logged to {PROMOTION_LOG_PATH.name}: batch={BATCH_LABEL!r}, "
        f"{len(newly_promoted)} promoted key(s) in {entry['groupCount']} group(s), "
        f"{len(newly_rejected)} rejected key(s)"
    )

## B6 — Guarantee the three floor conflicts

Check which floor conflicts are already covered by promoted (`trust_tier=1`) rows.
Extraction-preferred: only hand-add the ones missing below.

In [26]:
def floor_conflict_covered(subject: str, claim_type: str, min_distinct_values: int = 2) -> bool:
    promoted = [
        c for c in candidates
        if c["trust_tier"] == 1
        and c["subject_name"].strip().lower() == subject.lower()
        and c["claim_type"] == claim_type
    ]
    return len({c["claim_value"] for c in promoted}) >= min_distinct_values


coverage = {
    "Aphrodite parentage": floor_conflict_covered("Aphrodite", "parentage"),
    "Io parentage": floor_conflict_covered("Io", "parentage"),
    "Achilles death": floor_conflict_covered("Achilles", "death"),
}
for name, covered in coverage.items():
    print(f"{name}: {'covered' if covered else 'MISSING -- hand-add below'}")

Aphrodite parentage: covered
Io parentage: covered
Achilles death: covered


## Hand-add any floor conflict extraction missed

**Keep the `claim_type` canonical** (per DEV-018/DEV-020: Achilles' is `death`, not
`slaying`/`death_manner`) so a hand-added row groups with any extracted counterpart and
the exact-match `ConflictLookup` returns both. Run the normalize check first to
confirm the value you're about to use.

In [ ]:
import psycopg2

import config
from extraction.claim_type_normalizer import load_alias_map, normalize

conn = psycopg2.connect(
    host=config.POSTGRES_HOST, port=config.POSTGRES_PORT, dbname=config.POSTGRES_DB,
    user=config.POSTGRES_USER, password=config.POSTGRES_PASSWORD,
)
alias_map = load_alias_map(conn)
conn.close()

print(normalize(alias_map, "slaying"))    # -> "death"
print(normalize(alias_map, "parent_of"))  # -> "parentage"

In [ ]:
# Edit and run only for floor conflicts flagged MISSING above. Example shown for Io
# (Inachus vs Piren, per Apollodorus -- confirm the real passage_ref before running).
hand_added = [
    {
        "subject_name": "Io",
        "claim_type": "parentage",
        "claim_value": "daughter of Inachus",
        "source_id": "apollodorus-bibliotheca",
        "passage_ref": "2.1.3",
        "trust_tier": 1,
    },
    {
        "subject_name": "Io",
        "claim_type": "parentage",
        "claim_value": "daughter of Piren",
        "source_id": "apollodorus-bibliotheca",
        "passage_ref": "2.1.3",
        "trust_tier": 1,
    },
]

candidates.extend(hand_added)
CANDIDATES_PATH.write_text(json.dumps(candidates, indent=2, ensure_ascii=False), encoding="utf-8")
print(f"{len(hand_added)} rows hand-added; {len(candidates)} total rows now in the file")

## Final check

Re-run the B6 coverage cell above after any hand-adds — all three floor conflicts
must show `covered` before handing this off for Track C (`V12`).